### PR TITLE
modem: small improvements

### PR DESCRIPTION
**Pull request by tomi-font**
Changes include:
- Improving handling of frames for unconfigured DLCIs.
- Adding logs and outputting error codes in failure cases.
- Adding warnings to not reuse buffers.
- Fixing comments.

### CODE REVIEW

File: subsys/modem/modem_cmux.c [hunk 1]
```diff
@@ -876,12 +877,16 @@ static void modem_cmux_transmit_handler(struct k_work *item)
 		ret = modem_pipe_transmit(cmux->pipe, reserved, reserved_size);
 		if (ret < 0) {
 			ring_buf_get_finish(&cmux->transmit_rb, 0);
+			if (cmux->pipe->state != MODEM_PIPE_STATE_CLOSED) {
```
Comment (bjarki-trackunit) [inline subsys/modem/modem_cmux.c:880]: The modem_pipe_transmit API returns `-EPERM` if the pipe is closed, I can see this documentation is missing from the `modem_pipe.h`, it is only mentioned for the receive function... could you add this documentation, and check the return value instead of directly reading the internal context of the pipe? The pipe context is private, and should not be accessed directly :)

**Review by bjarki-andreasen — CHANGES_REQUESTED** [review]
Looks good overall, just one requested change (and the CI error)

File: subsys/modem/modem_cmux.c (hunk 1 above)
Comment (tomi-font) [reply subsys/modem/modem_cmux.c:881]: Okay, did as you suggested. However how about the hypothetical case where `pipe->api->transmit` (or `receive`, for `modem_pipe_receive()` would return `-EPERM`?

**Review by bjarki-andreasen — DISMISSED** [review]
Nice!

**Review by fabiobaltieri — DISMISSED** [review]

**Review by rerickson1 — DISMISSED** [review]

**Review by bbilas — DISMISSED** [review]

File: subsys/modem/backends/modem_backend_uart_async.c [hunk 2]
```diff
@@ -185,8 +185,8 @@ static int modem_backend_uart_async_transmit(void *data, const uint8_t *buf, siz
 	ret = uart_tx(backend->uart, backend->async.transmit_buf, bytes_to_transmit,
 		      CONFIG_MODEM_BACKEND_UART_ASYNC_TRANSMIT_TIMEOUT_MS * 1000L);
 
-	if (ret < 0) {
-		LOG_WRN("Failed to start async transmit");
+	if (ret) {
```
Comment (ycsin) [inline subsys/modem/backends/modem_backend_uart_async.c:188]: why this change?

Comment (ycsin) [reply subsys/modem/backends/modem_backend_uart_async.c:188]: and negative value if it isn't, L188 should be kept as-is, or change it to `ret != 0` if you really want to (explicit comparison is better, and it produces boolean result directly)

File: subsys/modem/modem_cmux.c [hunk 3]
```diff
@@ -830,6 +829,9 @@ static void modem_cmux_receive_handler(struct k_work *item)
 	/* Receive data from pipe */
 	ret = modem_pipe_receive(cmux->pipe, buf, sizeof(buf));
 	if (ret < 1) {
+		if (ret < 0) {
```
Comment (ycsin) [inline subsys/modem/modem_cmux.c:832]: something lesser than 1 must be lesser than 0

**Review by ycsin — DISMISSED** [review]

File: subsys/modem/modem_cmux.c (hunk 3 above)
Comment (tomi-font) [reply subsys/modem/modem_cmux.c:832]: Yes, so here this condition is equivalent to `ret != 0`. This is because a return value of 0 is not an error, thus we avoid logging an error in that case.

Comment (ycsin) [reply subsys/modem/modem_cmux.c:832]: > Yes, so here this condition is equivalent to `ret != 0`. This is because a return value of 0 is not an error, thus we avoid logging an error in that case.

ah sorry, my mind was reading `-1` instead of `1`

File: include/zephyr/modem/cmux.h [hunk 4]
```diff
@@ -195,7 +195,7 @@ void modem_cmux_init(struct modem_cmux *cmux, const struct modem_cmux_config *co
 struct modem_cmux_dlci_config {
 	/** DLCI channel address */
 	uint8_t dlci_address;
-	/** Receive buffer used by pipe */
+	/** Receive buffer used by pipe; must not be reused anywhere. */
```
Comment (ycsin) [inline include/zephyr/modem/cmux.h:198]: question: there are a few `receive_buf` in this file, why only mention `"must not be reused anywhere."` here? is there any receive buffer that can be reused and this one is an exception?

File: include/zephyr/modem/pipe.h [hunk 5]
```diff
@@ -130,25 +130,27 @@ void modem_pipe_attach(struct modem_pipe *pipe, modem_pipe_api_callback callback
...
- * @param buf Destination for reveived data
- * @param size Capacity of destination for recevied data
+ * @param buf Destination for received data; must not be already in use in a modem module.
+ * @param size Capacity of destination for received data
  *
- * @return Number of bytes received from pipe if any
- * @return -EPERM if pipe is closed
- * @return -errno code on error
+ * @retval Number of bytes received from pipe
+ * @retval -EPERM if pipe is closed
```
Comment (ycsin) [inline include/zephyr/modem/pipe.h:152]: I'd probably follow what's been used in the function desc already and keep the line change minimal, `@return` and `@retval` are use interchangeably in the entire zephyr source with no preference over one another.

File: include/zephyr/modem/cmux.h (hunk 4 above)
Comment (tomi-font) [reply include/zephyr/modem/cmux.h:198]: No, there is no receive buffer that can be reused. I added the comment here because I managed to trap myself with this buffer, so I though I'd add a warning to it.
When using CMUX, the DLCIs kind of are the point of communication between the user and the CMUX module. Not so much the other buffers.
Though anyway the most important place is `modem_pipe_receive()`, as that's how all the receiving takes place.

File: include/zephyr/modem/pipe.h (hunk 5 above)
Comment (tomi-font) [reply include/zephyr/modem/pipe.h:152]: I made the change from `@return` to `@retval` in order to make the choice uniform in this header.

File: include/zephyr/modem/cmux.h (hunk 4 above)
Comment (ycsin) [reply include/zephyr/modem/cmux.h:198]: > No, there is no receive buffer that can be reused.

In that case wouldn't this exception be rather confusing? i.e. now in the documentation you will have receive buffers and receive buffer that must not be reused anywhere

Comment (tomi-font) [reply include/zephyr/modem/cmux.h:198]: Could be. Although no buffer is specified as being reusable, but I get your point.
I could change the comment to `make sure not to reuse it anywhere`; what do you think?

**Comment by tomi-font** [comment]
ping @bjarki-trackunit @fabiobaltieri @rerickson1 @bbilas

**Review by bjarki-andreasen — DISMISSED** [review]

File: include/zephyr/modem/cmux.h (hunk 4 above)
Comment (ycsin) [reply include/zephyr/modem/cmux.h:198]: IMO this is simply about consistency, the fact that you only change this line is pretty random - you happened to have mistakenly shared a buffer for this type of struct which caused you some troubles - it probably could have been any other structs.

I'm not sure if anyone else would assume that a buffer can be shared among this kind of config struct, but one might start to wonder if (s)he saw that just one struct noted that its `receive_buf` shouldn't be reused, and the buffer of the same nature exists in a few other structs in this header.

tl;dr: probably not change this line, or change all of the lines, for consistency?

Comment (tomi-font) [reply include/zephyr/modem/cmux.h]: Alright. Removed it.

File: subsys/modem/backends/modem_backend_uart_async.c [hunk 6]
```diff
@@ -185,8 +185,8 @@ static int modem_backend_uart_async_transmit(void *data, const uint8_t *buf, siz
 	ret = uart_tx(backend->uart, backend->async.transmit_buf, bytes_to_transmit,
 		      CONFIG_MODEM_BACKEND_UART_ASYNC_TRANSMIT_TIMEOUT_MS * 1000L);
 
-	if (ret < 0) {
-		LOG_WRN("Failed to start async transmit");
+	if (ret != 0) {
+		LOG_ERR("Failed to start async transmit. (%d)", ret);
```
Comment (ycsin) [inline subsys/modem/backends/modem_backend_uart_async.c:189]: it would be nice if such error messages are consistent so that the same buffer can be reused, which would result in smaller ROM size which I'm sure that an embedded device would appreciate, i.e.:

```c
LOG_ERR("%d: failed to %s %d bytes", ret, "async transmit", bytes_to_transmit);
```

or 

```c
LOG_ERR("%d: %s failed", ret, "async transmit");
```

File: subsys/modem/modem_cmux.c [hunk 7]
```diff
@@ -830,6 +829,9 @@ static void modem_cmux_receive_handler(struct k_work *item)
...
+			LOG_ERR("Pipe reading error: %d", -ret);
```
Comment (ycsin) [inline subsys/modem/modem_cmux.c:833]: i.e.:

```c
LOG_ERR("%d: failed to %s %d bytes", ret, "read pipe", sizeof(buf));
```

or 

```c
LOG_ERR("%d: %s failed", ret, "pipe_receive");
```

File: subsys/modem/modem_cmux.c [hunk 8]
```diff
@@ -876,12 +878,16 @@ static void modem_cmux_transmit_handler(struct k_work *item)
 		ret = modem_pipe_transmit(cmux->pipe, reserved, reserved_size);
 		if (ret < 0) {
 			ring_buf_get_finish(&cmux->transmit_rb, 0);
+			if (ret != -EPERM) {
+				LOG_ERR("Failed to transmit %u bytes. (%d)", reserved_size, ret);
```
Comment (ycsin) [inline subsys/modem/modem_cmux.c:882]: i.e.:

```c
LOG_ERR("%d: failed to %s %d bytes", ret, "transmit", reserved_size);
```

or 

```c
LOG_ERR("%d: %s failed", ret, "pipe_transmit");
```

**Comment by fabiobaltieri** [comment]
> The documentation build is somehow broken, and a rebase didn't help.

`The self-hosted runner: zephyr-runner-linux-x64-4xlarge-deployment-g2tcl-xtbmv lost communication with the server.` runner problem, hit retry, should work eventually

**Review by ycsin — COMMENTED** [review]
I think it would be nice if the commit can be split since they target different matters, but all these are just suggestions and non-blocking

**Comment by ycsin** [comment]
the doc build fails in #67020 as well, I reran the job there and it didn't help

**Comment by fabiobaltieri** [comment]
> the doc build fails in #67020 as well, I reran the job there and it didn't help

This one is happy now, the other one looks like an actual problem problem, let's continue on that PR thread.

**Review by rerickson1 — APPROVED** [review]

**Review by ycsin — APPROVED** [review]

**Comment by tomi-font** [comment]
ping @bjarki-trackunit

**Review by bjarki-andreasen — APPROVED** [review]
